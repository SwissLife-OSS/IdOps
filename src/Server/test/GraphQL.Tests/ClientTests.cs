using System.Threading.Tasks;
using HotChocolate;
using HotChocolate.Execution;
using IdOps.Authorization;
using IdOps.Security;
using Snapshooter.Xunit;
using Squadron;
using Xunit;

namespace IdOps.GraphQL.Tests
{
    [Collection(TestCollectionNames.GraphQL)]
    public class ClientTests : TestHelper
    {
        public ClientTests(MongoResource resource) : base(resource)
        {
        }

        [Theory]
        [InlineData(Roles.Admin)]
        [InlineData(Roles.Edit)]
        public async Task ValidRequest(string role)
        {
            // arrange
            ITestRequestBuilder requestBuilder = TestRequestBuilder.New()
                .AddExecutor(await CreateSchema())
                .AddRequestFromFile("CreateClient")
                .AddUser()
                .AddRole(role);
            
            await TestDataBuilder
                .New(Services)
                .SetupTenant()
                .ExecuteAsync();

            // act
            IExecutionResult result = await requestBuilder.ExecuteAsync();

            // assert
            result.ToJson().MatchSnapshot(x => x.IgnoreFields("**.id"));
        }

        [Theory]
        [InlineData(Roles.Read)]
        [InlineData(null)]
        public async Task CreateClient_ByUserWithoutPermission_IsDenied(string? role)
        {
            // arrange
            ITestRequestBuilder requestBuilder = TestRequestBuilder.New()
                .AddExecutor(await CreateSchema())
                .AddRequestFromFile("CreateClient")
                .AddUser()
                .AddRole(role);

            // act
            IExecutionResult result = await requestBuilder.ExecuteAsync();

            // assert
            result.ToJson()
                .MatchSnapshot();
        }

        [Fact]
        public async Task CreateClient_NotAuthenticated()
        {
            // arrange
            ITestRequestBuilder requestBuilder = TestRequestBuilder.New()
                .AddExecutor(await CreateSchema())
                .AddRequestFromFile("CreateClient")
                .SetAuthenticated(false);

            // act
            IExecutionResult result = await requestBuilder.ExecuteAsync();

            // assert
            result.ToJson()
                .MatchSnapshot();
        }
    }
}
