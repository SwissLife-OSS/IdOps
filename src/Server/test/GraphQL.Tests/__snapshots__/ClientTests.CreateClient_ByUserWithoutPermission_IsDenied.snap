{
  "errors": [
    {
      "message": "The current user is not authorized to access this resource.",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "createClient"
      ],
      "extensions": {
        "code": "AUTH_NOT_AUTHENTICATED"
      }
    }
  ]
}
